app/controllers/lottery_controller.rb

```diff
@@ -12,64 +12,103 @@ def action_allowed?
+    teams_not_signing_up = teams.reject {|team| SignedUpTeam.where(team_id: team.id, is_waitlisted: 0).any? }
+    teams_not_signing_up.each do |team|
+      bids = [0]
+      sign_up_topics.each do |topic|
+        bid_record = Bid.find_by(team_id: team.id, topic_id: topic.id)
+        bids << bid_record.priority if bid_record.try(:priority)
+      end
+      # Grab student id and list of bids
+      team.users.each {|user| user_bidding_info << {pid: user.id, ranks: bids} } if bids.uniq != [0]
+    end
+    user_bidding_info
+  end
+
+  # Generate team bidding infomation hash based on newly-created teams
+  # Structure of team_bidding_info variable: [{team_id1, bids_1}, {team_id2, bids_2}]
+  def construct_team_bidding_info(unassigned_teams, sign_up_topics)
+    team_bidding_info = []
+    unassigned_teams.each do |team|
+      topic_bids = []
+      sign_up_topics.each do |topic|
+        bid = Bid.find_by(team_id: team.id, topic_id: topic.id)
+        topic_bids << {topic_id: topic.id, priority: bid.priority} if bid
+      end
+      topic_bids.sort! {|bid| bid[:priority] }
+      team_bidding_info << {team_id: team.id, bids: topic_bids}
+    end
+    team_bidding_info
+  end
+
+  def create_new_teams_for_bidding_response(teams, assignment, user_bidding_info)
     teams.each do |user_ids|
       # Create new team and team node
       new_team = AssignmentTeam.create(name: 'Team_' + rand(10_000).to_s, parent_id: assignment.id)
       team_node = TeamNode.create(parent_id: assignment.id, node_object_id: new_team.id)
+
       user_ids.each do |user_id|
-        # Destroy current team_user and team_user node if exists
-        TeamsUser.where(user_id: user_id).each do |team_user|
-          next unless team_user.team.parent_id == assignment.id
-          team_user.team_user_node.destroy rescue nil
-          team_user.destroy rescue nil
-          break
-        end
+        remove_current_user(assignment.id, user_id)
+
         # Create new team_user and team_user node
         new_team_user = TeamsUser.create(user_id: user_id, team_id: new_team.id)
         TeamUserNode.create(parent_id: team_node.id, node_object_id: new_team_user.id)
+
         # Create new bids for team based on `bidding_data` variable for each team member
         # Currently, it is possible (already proved by db records) that
         # some team has multiple 1st priority, multiply 2nd priority, ....
         # these multiple identical priorities come from different previous teams
         # [Future work]: we need to find a better way to merge bids that came from different previous teams
-        merge_bids_from_different_previous_teams(assignment, new_team.id, user_ids, priority_info)
+        merge_bids_from_different_previous_teams(assignment, new_team.id, user_ids, user_bidding_info)
       end
     end
-    # Remove empty teams
+  end
+
+  # Destroy current team_user and team_user node if exists
+  def remove_current_user(assignment_id, user_id)
```

Comment:
I think a better name for this method would be something like remove_user_from_previous_team to clarify that the user isn't getting removed but just disassociated from their old team.